src/core/thread/context.d

```diff
@@ -0,0 +1,312 @@
+/**
+ * Every thread / fiber needs stack context data structure.
+ * thread.context decouples this info into the StackContext struct.
+ * Also, Thread / Fiber have some common functionality which is provided
+ * as a super class for both in StackContextExecutor.
+ * Finally, GlobalStackContext contains commonly used global data.
+ *
+ * Copyright: Copyright Sean Kelly 2005 - 2012.
+ * License: Distributed under the
+ *      $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost Software License 1.0).
+ *    (See accompanying file LICENSE)
+ * Authors:   Sean Kelly, Walter Bright, Alex Rønne Petersen, Martin Nowak
+ * Source:    $(DRUNTIMESRC core/thread/osthread.d)
+ */
+
+module core.thread.context;
+
+struct StackContext
+{
+    // bottom, top of stack
+    void* bstack, tstack;
+
+    /// Slot for the EH implementation to keep some state for each stack
+    /// (will be necessary for exception chaining, etc.). Opaque as far as
+    /// we are concerned here.
+    void* ehContext;
+
+    StackContext* within, next, prev;
+}
+
+/**
+A class that represents a thread of execution that manages a stack.
+This serves primarily as a superclass for Thread and Fiber.
+*/
+class StackContextExecutor
+{
+    //
+    // The type of routine passed on thread/fiber construction.
+    //
+    enum Call
+    {
+        NO,
+        FN,
+        DG
+    }
+
+    // Common standard data for Thread / Fiber.
+    Call                m_call = Call.NO;
+    union
+    {
+        void function() m_fn;
+        void delegate() m_dg;
```

Comment:
It would be nice to package this up in a way that deduplicates the amount of work in the constructors of Thread and Fiber. something like 

```d
static struct Entrypoint
{
    enum Call {  NO, FN, DG }
    Call                m_call = Call.NO;
    union
     {
         void function() m_fn;
         void delegate() m_dg;
    }
    void opAssign(void function() fn) @trusted
    {
         m_fn = fn;
         m_call = Call.FN; 
    }

    // ditto for delegate

    void opCall()
    {
        run();
    }
}
```

Reply:
Ok but it seems kind of risky at this point. I'll leave for the end.

Reply:
Oh, thats definitely for another PR, I just wanted to make a note of it while I remembered.